Wait for "stopped" status in StopService

myles-keough opened the pull request:
I've changed this to poll the service every 1 second, so we should be able to return shortly after the service ends. This is repeated up to 90 times.

Changed some comments to remove ambiguity.

jamesodoherty-huntress reviewed (approved):
Nice work. This behaves much better with the calling site logic:

https://github.com/huntresslabs/win-service-updater/blob/b8a3565ad8e505cf019dc9875df4f0523c091093/updater/update.go#L166-L172

However, we still might want to revisit the call site where it's possible in a few spots for hUpdate to bail after stopping services and before finishing starting services.

For example, even if everything is successfully stopped, if we encounter a single error while starting any of the services, we fail to start the remaining services:

https://github.com/huntresslabs/win-service-updater/blob/b8a3565ad8e505cf019dc9875df4f0523c091093/updater/update.go#L176-L192

So we should also think about updating the `InstallUpdate` function to unconditionally try starting all services after stopping services, and to try restarting any services that have already been stopped if we bail early on stopping all the services.

tramzee commented on updater/servicecontrol_windows.go:
```diff
@@ -62,6 +62,37 @@ func IsServiceRunning(serviceName string) (bool, error) {
+		return true, nil
+	}
+
+	if status.State != svc.Stopped {
+		return false, nil
+	}
+
+	return true, nil
```
Maybe: `return status.State == svc.Stopped, nil`?

tramzee commented on updater/servicecontrol_windows.go:
```diff
@@ -110,28 +141,21 @@ func StopService(serviceName string) error {
 	}
 
+	status, _ := s.Query()
 	if status.State != svc.Stopped {
-		err = fmt.Errorf("'%s' did not stop; status: %+v", serviceName, status)
+		err = fmt.Errorf("'%s' did not stop in time; status: %+v", serviceName, status)
 		return err
 	}
```
If we get to the end of the loop we pretty much know that the service is not stopped so we can just return an error here.

Just as a by-the-way, what is the difference between the `s.Query` used here and the `s.Control(svc.Interrogate)` used above?

myles-keough replied:
It looks like `s.Query` will call `QueryServiceStatusEx` internally and `s.Control(svc.Interrogate)` will actually send an event to the service control handler and wait for a response. I'm going to modify this to use query only.

tramzee commented on updater/servicecontrol_windows.go:
```diff
@@ -110,28 +141,21 @@ func StopService(serviceName string) error {
 	for i := 0; i < retries; i++ {
-		running, _ := IsServiceRunning(serviceName)
+		time.Sleep(1 * time.Second)
+		stopped, err := IsServiceStopped(serviceName)
```
Since we have the service open here do we want to pass that in to `IsServicedStopped` instead of opening the same service again inside `IsServiceStopped`? I expect this is ok but I do worry...

myles-keough commented:
@jamesodoherty-huntress If there are any errors returned from the InstallUpdate call, all of the files will be rolled back and services started again. I do agree that in the future these should be coupled better within InstallUpdate. It doesn't make much sense to require a calling function clean up another function's mess.

https://github.com/huntresslabs/win-service-updater/blob/b8a3565ad8e505cf019dc9875df4f0523c091093/updater/handler.go#L199-L210

myles-keough commented:
I've removed `IsServiceStopped` and replaced it with `Query` calls to our current service handle. I've also bumped the retries to allow for up to 130 seconds. I've discovered that Windows will allow a maximum of 125 seconds for a service to stop before killing the process - though in our services this should be finished within 30 seconds since I don't think we ever use `RequestAdditionalTime`.

jamesodoherty-huntress commented on updater/servicecontrol_windows.go:
```diff
@@ -141,23 +110,18 @@ func StopService(serviceName string) error {
 	}
 
-	var retries int = 90
+	var retries int = 130
```
It might be good to add a comment with the URL to the Services snap-in deadline so we know where the 130 seconds comes from.
